### PR TITLE
Changing resolution in the access coupled model

### PR DESCRIPTION
**Pull request by ScottWales**

### CODE REVIEW

**Comment by ScottWales** [comment]
Notebook is at https://github.com/coecms/coecms.github.io/blob/79bd72bc3a1a9a0d6a64ab58e1044ef4ec6c46ff/notebooks/2019-03-01-coupling-resolution.ipynb

File: _includes/notebooks/2019-03-01-coupling-resolution.html [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+</div>
+<div class="inner_cell">
+<div class="text_cell_render border-box-sizing rendered_html">
+<h2 id="Regridding-weights">Regridding weights<a class="anchor-link" href="#Regridding-weights">&#182;</a></h2><p>Oasis requires regridding weights in SCRIP format to convert a field from one model to another. Oasis can generate these itself from SCRIP descriptions of the source and target grid, or they can be generated using <a href="http://www.earthsystemmodeling.org/esmf_releases/public/ESMF_7_1_0r/ESMF_refdoc/node3.html#SECTION03020000000000000000">ESMF_RegridWeightGen</a> and the ESMF weights converted to SCRIP format.</p>
+<p>The most important thing here is that the masks are consistent between all the models. If they are not you will see discontinuity in the surface fields around coastlines.</p>
+<h3 id="UM-Grids">UM Grids<a class="anchor-link" href="#UM-Grids">&#182;</a></h3><p>The UM mask is defined by two files - the land mask, which is 1 everywhere there is greater than 0% land in a grid cell, and the land fraction, which is the fraction of land in a grid cell. These fields are only defined on the T grid.</p>
```
Comment (RavDanny) [inline _includes/notebooks/2019-03-01-coupling-resolution.html]: I don't know what a T grid is. You might want to explain or provide a link.

File: _includes/notebooks/2019-03-01-coupling-resolution.html [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+<p>The coupling mask for the UM T grid should be 1 everywhere the land fraction field is less than 1, in other words everywhere with greater than 0% ocean in a cell.</p>
+<p>For the U and V fields I used no mask, since we don't want to conservatively interpolate vector fields.</p>
```
Comment (RavDanny) [inline _includes/notebooks/2019-03-01-coupling-resolution.html]: I don't know what the U and V fields are.

File: _includes/notebooks/2019-03-01-coupling-resolution.html [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+<p>The most important thing here is that the masks are consistent between all the models. If they are not you will see discontinuity in the surface fields around coastlines.</p>
+<h3 id="UM-Grids">UM Grids<a class="anchor-link" href="#UM-Grids">&#182;</a></h3><p>The UM mask is defined by two files - the land mask, which is 1 everywhere there is greater than 0% land in a grid cell, and the land fraction, which is the fraction of land in a grid cell. These fields are only defined on the T grid.</p>
+<p>The coupling mask for the UM T grid should be 1 everywhere the land fraction field is less than 1, in other words everywhere with greater than 0% ocean in a cell.</p>
```
Comment (RavDanny) [inline _includes/notebooks/2019-03-01-coupling-resolution.html]: Need an explanation of what a coupling mask is.

File: _includes/notebooks/2019-03-01-coupling-resolution.html [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+<p>For the U and V fields I used no mask, since we don't want to conservatively interpolate vector fields.</p>
+<p>Since the UM uses regular lat-lon grids you can create a GRIDSPEC grid description to feed into ESMF.</p>
```
Comment (RavDanny) [inline _includes/notebooks/2019-03-01-coupling-resolution.html]: Same with GRIDSPEC. Do you want to provide a link to explain it?
Can you explain what ESMF is?

File: _includes/notebooks/2019-03-01-coupling-resolution.html [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+<h3 id="MOM/CICE-Grids">MOM/CICE Grids<a class="anchor-link" href="#MOM/CICE-Grids">&#182;</a></h3><p>The MOM and CICE masks should be equal to each other. The CICE mask is found in <code>$ICE/INPUT/kmt.nc</code> as variable <code>kmt</code>, the MOM mask in <code>$OCN/INPUT/grid_spec.nc</code> as variable <code>wet</code>.</p>
+<p>The coupling mask for the MOM/CICE T grid should be equal to <code>wet</code> from <code>grid_spec.nc</code>.</p>
```
Comment (RavDanny) [inline _includes/notebooks/2019-03-01-coupling-resolution.html:23]: As above, need an explanation what a coupling mask is.

File: _includes/notebooks/2019-03-01-coupling-resolution.html [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+<p>All of the fields needed to create a SCRIP description of a MOM grid may be found in <code>grid_spec.nc</code></p>
+<h3 id="Masks-and-Conservative-Regridding">Masks and Conservative Regridding<a class="anchor-link" href="#Masks-and-Conservative-Regridding">&#182;</a></h3><p>For conservative regridding to work, the masks of the two models must be consistent as well - there can't be areas where the ocean model thinks it's land and the atmosphere model thinks it's ocean.</p>
```
Comment (RavDanny) [inline _includes/notebooks/2019-03-01-coupling-resolution.html]: What's **conservative** regridding?
Later you refer to patch regridding. Do you want to explain that there are 2 types of regridding (at least?) and how they work?

File: _includes/notebooks/2019-03-01-coupling-resolution.html [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+<p>An example point where this can happen is the ice shelves around Antarctica. The ocean model's grid doesn't go all the way to the south pole, so it misses the southernmost part of the Ross ice shelf. The atmosphere grid does reach the pole however, and its default land mask assumes the ice shelf is ocean. There can be other areas where this causes conflicts as well - for instance the black sea is not included in the MOM default mask.</p>
+<p>To make sure that the grids are consistent you can do a conservative regridding of the mask from one grid to the other. The highest resoultion grid should be the source, in my case I used the ocean mask. I then conservatively regridded the <code>wet</code> field from MOM to the UM grid. It's important to not use masks for this interpolation, also since the MOM grid doesn't cover the entire globe I turned on the ESMF '--ignore_unmapped' flag. This gave me an ocean fraction field on the UM grid. The ocean fraction can then be straightforwardly converted to a land fraction and land mask field to be input to the UM.</p>
```
Comment (RavDanny) [inline _includes/notebooks/2019-03-01-coupling-resolution.html]: * Is that saying: "you can do a conservative regridding of the mask from one grid to the other using ESMF"?
* typo "resoultion"
* "It's important to not use masks for this interpolation" - not sure what this means. Do you want to explain why?
* "This gave me an ocean fraction field on the UM grid." - Are you saying: "Running the regrid gave me..."?

File: _includes/notebooks/2019-03-01-coupling-resolution.html [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+<p>Conservative regridding is used for scalar fields, patch regridding is used for vector field components.</p>
+
+</div>
+</div>
+</div>
+<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
+</div>
+<div class="inner_cell">
+<div class="text_cell_render border-box-sizing rendered_html">
+<h2 id="UM-Ancillary-interpolation">UM Ancillary interpolation<a class="anchor-link" href="#UM-Ancillary-interpolation">&#182;</a></h2><p>Since the UM is using a new resolution and mask the various ancillary datasets also need to be changed. For the most part files from the Met Office archive at the target resolution were used. For fields like vegetation fraction which are only defined on land points I 'de-masked' the anciallary field, by performing a nearest grid point interpolation from the masked source field to the same grid without a mask. It's important to use nearest grid point so that fields like the vegetation fractions correctly add up - some interpolation types also can result in negative values at some points which crashes the model.</p>
```
Comment (RavDanny) [inline _includes/notebooks/2019-03-01-coupling-resolution.html]: typo "anciallary"
"For fields like vegetation fraction which are only defined on land points" as opposed to lat/lon? Either way, can you explain why you needed to do what you did?

File: _includes/notebooks/2019-03-01-coupling-resolution.html [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+<p>The aerosol climatology files had to be vertically interpolated since we had also reduced the number of vertical levels in the model run. For this I used the <a href="https://github.com/SciTools-incubator/python-stratify">stratify</a> python library. The vertical levels used by the UM are hybrid height levels - they are terrain following up to a certain height, then above that they are constant height levels.</p>
```
Comment (RavDanny) [inline _includes/notebooks/2019-03-01-coupling-resolution.html]: Do you have to explain what "terrain" means? Does it mean variable height levels?

File: _includes/notebooks/2019-03-01-coupling-resolution.html [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+<p>To convert from the source set of hybrid heights to the target hybrid heights I first converted both to true heights, using the formula from <a href="https://code.metoffice.gov.uk/doc/um/latest/papers/umdp_F03.pdf">UMDB F03</a>. I then used stratify to convert the aerosol fields on true source heights to the true target heights.</p>
+<p>To modify the UM ancillary files I used the <a href="https://code.metoffice.gov.uk/doc/um/mule/2018.07.1/">mule</a> library as well as <a href="https://pandas.pydata.org/">pandas</a> to group the 2d field slices in the file into the full 4d field</p>
```
Comment (RavDanny) [inline _includes/notebooks/2019-03-01-coupling-resolution.html]: Do you want to explain what 2d field slices and the full 4d field are?

File: _includes/notebooks/2019-03-01-coupling-resolution.html [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+    <span class="k">return</span> <span class="n">Z</span>
+</pre></div>
+
+</div>
+</div>
+</div>
+
+</div>
+
```
Comment (RavDanny) [inline _includes/notebooks/2019-03-01-coupling-resolution.html:132]: Do you want to have a closing line not to hit a sudden halt? E.g. "And then it's done!"

**Review by RavDanny — COMMENTED** [review]
Interesting. Not sure how much assumed knowledge should be expected here. A lot was unfamiliar to me.
I should have made my comments on the .pnyb file. Anyway, it's done now.

**Comment by ScottWales** [comment]
Thanks for the review @RavDanny, appreciate all the suggestions

**Pull request merged**